### DISCUSSION TITLE
SonoOff 1 Channel - Adding LED Status Input Sensor

### DISCUSSION THREAD

**Discussion by JayR33** [arendst/Tasmota#17914] [Q&A]
Hi All,
This may be a simple question in comparison to some of the questions being asked.
I am using a SonoOff 1 relay to turn a device on and off remotely.  The device has an onboard status LED which activates at a set noise level which I would like to monitor whether it is on or off (detected a high noise level).  The voltage on the LED pins is floating when off, and 2v when on.
I understand that GPIO17 can be used as an ADC input with a maximum input voltage of 1v.  I could make up a voltage divider and wire that to GPIO17?
Is there a simpler way monitoring the status of the LED?
I read that the Switch component/setting is considered a digital input (just on or off).  Is the input of the other GPIO's (1,3,4,5,14) also 1v maximum, therefore I'd need a voltage divider anyway?
Thanks!

**Comment by barbudor** [comment]
There might be considerations about common ground
Without a more detailed schematic, I couldn't guarantee that this hack wouldn't end up in magic smoke

As the device drives a led, why not adding an optocoupler?
Connected the opto's led in parallel of the existing led. The brightness of the leds may reduce as current will be shared among the 2 leds
Then use the transistor side to drive a gpio
Preferably a standard gpio
With any gpio but 15, connect emitter to gnd and collector to gpio. When the led is off, gpio will be high. When the led is on, gpio will be low. Use the proper switchmode
With gpio 15 which has a on board pull down, connect emitter to gpio and collector to VCC. Logic will be inverted.

**Reply by JayR33** [reply]
Ah yes, an opto is a good idea. And I think I have a few here I can try.

The challenging part will be configuring Tasmota to be correct,  but I'll see how I go.
Thanks again.

**Reply by barbudor** [reply]
What is exactly "Sonoff 1 Relay" ? A Sonoff Basic ? A Sonoff SV ?

I can provide a suggested wiring and configuration once I know the device

**Reply by JayR33** [reply]
Thanks Barbudor, muchly appreciated!

The module I'm using is the PSF-B01 board listed on this page - https://templates.blakadder.com/sonoff_1channel.html

I found a cheap relay module board that has an optocoupler onboard that I can use, I'll cut/graft that into place.  I knew I'd find a use for it one day...

Ideally I'd like to be able to send the on/off status of the LED over mqtt to Home Assistant.

Thanks again

**Reply by barbudor** [reply]
Ok, So here a quick diagram of what to achieve:

![image](https://user-images.githubusercontent.com/2996042/218264141-c2b5da34-6380-4a2a-b013-155bbd3a765e.png)

GPIO0 and GPIO2 must be high at boot
So better not to use them for that purpose because if the ESP was to restart while the led is ON, then the ESP wouldn't boot.

GPIO15 has a pull-down so a connection like the above would be with no effect. But if you were to connect the transistor output between GPIO15 and VCC, similar problem could occur as GPIO15 must be down at boot time.

Any other standard GPIO can be used.

In Configure -> Configure Module or Configure -> Configure Template, set the matching GPIO to `Switch` and the index to the 1st free index. If you don't yet have a Switch, it must be Switch 1

The next steps are depending what you want to achieve with that switch.
If you just want the status to be sent to MQTT, you can use `SwitchmodeX 16` so that the switch emits ON when the led is ON and OFF when the led is OFF (where X is the index of your switch)

**Reply by JayR33** [reply]
G'day Barbudor,

Thanks very much for the time to draw up the diagram and write out the explaination/config.  It helped imensly!!

Everything worked perfectly except I found I had to put a ~500k ohm resistor in series with the input of the optocoupler.  It seems the voltage threshold of the opto is lower than the LED, but using the resistor in series with the input of the opto fixed that.

While I vaugly remember reading something about GPIO pins 0,2 and 15, I hadn't thought about them.  You having off the top of your head expereince (and willing to share) has helped a lot.

Regarding the MQTT, I set GPIO5 to Switch 1 as I have no other Switches configured, the ran SwitchMode1 16 in the console and I can see in the console the  "Switch1":"ON" and "Switch1":"OFF" when the status changes - perfect.  But at 5min intervals I seem to get a ghost trigger.

The console reads

10:49:54.402 MQT: tele/piezo_switch/STATE = {"Time":"2023-02-12T10:49:54","Uptime":"0T00:10:09","UptimeSec":609,"Heap":19,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":23,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"ZH5","BSSId":"98:DA:C4:F1:31:EA","Channel":12,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:04"}}

10:49:54.410 MQT: tele/piezo_switch/SENSOR = {"Time":"2023-02-12T10:49:54","Switch1":"ON"}

10:54:54.413 MQT: tele/piezo_switch/STATE = {"Time":"2023-02-12T10:54:54","Uptime":"0T00:15:09","UptimeSec":909,"Heap":15,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"ZH5","BSSId":"98:DA:C4:F1:31:EA","Channel":12,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:04"}}

10:54:54.422 MQT: tele/piezo_switch/SENSOR = {"Time":"2023-02-12T10:54:54","Switch1":"ON"}

But at 10:49:54 and 10:54:54 the LED didn't trigger to the ON status on the device.  It seems the /STATE line is causing MQTT to send the ON message even when the device LED isn't on.

Have you seen this before, and any idea how to stop the false "ON" MQTT messages every 5mins?

**Comment by JayR33** [comment]
On a second thought - the MQTT message sent every 5mins is the status details of the Sonoff SV showing power on status, SSID of the conntected AP, RSSI and signal etc.  I wonder why when the status MQTT message is causing the Switch1 status to change from the OFF status to the ON status?

**Reply by barbudor** [reply]
Not sure to understand your point
The STATE message gives the status of the Relay (power) not of the switch
The SENSOR message gives the state of the switch

Depending on your configuration, those may be linked or not
Commands such as Switchmode, Setoption114 and rules change the behavior of switches and their possible attachement to the relay with the same index.

I briefly wrote about 'Switchmode 16' as one possible configuration but I don't know what you actually decide and configure

You can read that: https://tasmota.github.io/docs/Buttons-and-Switches/

And/or eventually explain exactly what you would like to happen from the switch to mqtt or actions on the relay or something else
